PCVSentinel

ditchfieldcaleb opened the pull request.

eswak commented on contracts/pcv/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.4;
+
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import "../../refs/CoreRef.sol";
+import "./IPCVSentinel.sol";
+import "../IPCVDeposit.sol";
+import "../../libs/CoreRefPauseableLib.sol";
+import "./IGuard.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
```
Can you sort ? 😅

eswak commented on contracts/pcv/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,173 @@
+        override
+        returns (address[] memory all) 
+    {
+        all = new address[](guards.length());
+
+        for (uint i = 0; i < guards.length(); i++) {
+            all[i] = guards.at(i);
+        }
+
+        return all;
```
guards.values() should be enough here ?

ditchfieldcaleb replied:
we use openzeppelin v3 not v4 so we don't have access to .values :(

Joeysantoro replied:
whaa really? Thats surprising

ditchfieldcaleb replied:
> whaa really? Thats surprising

yeah I spent a few hours once trying to fix it but that requires using *both* versions in our contract compilation process for anything that depends on core (since we use it in core!) and that became a nightmare

wouldn't recommend it

eswak commented on contracts/pcv/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,173 @@
+        address[] memory guardsWhoCanProtec = new address[](numGuardsCanProtec);
+
+        for (uint256 i=0; i<guards.length(); i++) {
+            if(IGuard(guards.at(i)).check()) {
```
you could cache these check() results in a boolean array to save external calls (I know it's not supposed to be called onchain, but...)

ditchfieldcaleb replied:
yeah if I want to use this onchain I'll definitely make an optimized version of this, makes it harder to read & audit though

ditchfieldcaleb replied:
wait nvm this isn't hard. good callout.

eswak commented on contracts/pcv/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,173 @@
+        bool needsProtec = IGuard(guard).check();
+
+        if (needsProtec) {
+            (address[] memory targets, bytes[] memory datas) = IGuard(guard).getProtecActions();
```
"data" has no plural

ditchfieldcaleb replied:
technically you are correct, but I add an s to help differentiate an array of data-objects from a single data object

Joeysantoro replied:
calldatas?

eswak commented on contracts/pcv/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,173 @@
+
+            for(uint256 i=0; i<targets.length; i++) {
+                require(targets[i] != address(this), "Nyeh!");
+                targets[i].call(datas[i]); // Unopinionated about reverts
```
What do you mean by "Unopinionated about reverts" ?

ditchfieldcaleb replied:
the pcv sentinel does not care if the actions it takes are reverting. adds complexity; the guard can handle that.

Joeysantoro replied:
I actually feel pretty strongly we should fail loud

eswak commented on test/unit/pcv/sentinel/PCVSentinel.test.ts:
```diff
@@ -0,0 +1,202 @@
+// Import if needed, just a helper.
+// const toBN = ethers.BigNumber.from;
+
+describe.only('PCV Sentinel', function () {
```
don't commit the .only ^^

ditchfieldcaleb replied:
oop will fix thanks

Joeysantoro commented on contracts/pcv/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,173 @@
+        override
+    {
+        for(uint256 i=0; i<whichGuards.length; i++) {
+            try this.protec(whichGuards[i]) {
```
Swallowing errors is a huge defi antipattern.

I think we are far more at risk of missing something here than being screwed because we failed loud. Need some pretty specific contrary evidence to change my mind

ditchfieldcaleb replied:
Require flawless execution of proective actions is an antipattern in security.
I agree that the risk of not knowing if something has failed is large, but I think the risk of not preventing a hack is much larger.

We could add in a bool that lets the caller decide though.

ElliotFriedman replied:
I agree with Joey here, we should not just ignore errors when we are trying to protec.

ditchfieldcaleb replied:
@Joeysantoro @ElliotFriedman neither of you have addressed the fact that the only way to hear about errors is events here.

> If someone calls into this contract from another contract and it reverts, we won't see those. We'll see the events if it doesn't revert though.

Therefore the best way to fail loudly is to *not* revert.

I added in a bool that lets the caller decide if it should revert or not.

ditchfieldcaleb replied:
(it's important to remember here that we won't be the only ones calling into this contract)

ElliotFriedman replied:
That is a case for only allowing all or nothing tx's.

Given that ethereum is a dark forest and a highly adversarial environment, we should not allow our contracts to halfway execute a transaction that is supposed to protect many of them at the same time.

ditchfieldcaleb replied:
> That is a case for only allowing all or nothing tx's.
> 
> Given that ethereum is a dark forest and a highly adversarial environment, we should not allow our contracts to halfway execute a transaction that is supposed to protect many of them at the same time.

Sorry I don't really know what you're saying here

Joeysantoro commented on contracts/pcv/sentinel/guards/examples/BalanceGuard.sol:
```diff
@@ -0,0 +1,33 @@
+        override 
+        returns (bool) 
+    {
+        if(address(this).balance % 2 == 0) return true;
```
assuming this is just a test?

ditchfieldcaleb replied:
yeah, it lets me send eth to it and change the trigger condition

ElliotFriedman commented on contracts/pcv/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,173 @@
+
+            for(uint256 i=0; i<targets.length; i++) {
+                require(targets[i] != address(this), "Nyeh!");
+                targets[i].call(datas[i]); // Unopinionated about reverts
```
you should emit an event here with the result of each call, true or false, the target address, and the returned data if any.

ditchfieldcaleb replied:
switched this to ensuring it doesn't fail. I really don't think we need an event for each individual sub-call though.

ElliotFriedman commented on contracts/pcv/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,173 @@
+
+            emit Protected(guard);
+        } else {
+            revert("No need to protec.");
```
no need to revert if there is nothing to protec

instead, have the function return true or false and return true if anything was protected and false otherwise

ditchfieldcaleb replied:
Reverting lets us bubble up errors and emit events when executing multiple protec actions

ElliotFriedman replied:
I'm not convinced this is the right decision.

If you pass in an array of guards to protec, and only half of them can be protec'd, why should we revert when the revert is ignored?

Also, any errors that try to bubble up get swallowed and are forgotten so this error string is literally pointless unless you try to call protec individually, and even then, it's much nicer to return a boolean than fail loudly.

If protecMany reverted on errors to protec, then I would be fine with this pattern.

ElliotFriedman commented on contracts/pcv/sentinel/guards/examples/BalanceGuard.sol:
```diff
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
```
this should go in the mocks folder

ditchfieldcaleb replied:
It's not a mock, it's an example

ElliotFriedman commented on contracts/pcv/sentinel/guards/examples/MultiActionGuard.sol:
```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
```
this should go in the mocks folder

ditchfieldcaleb replied:
(See above)

ElliotFriedman commented on contracts/pcv/sentinel/guards/examples/NoOpGuard.sol:
```diff
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
```
this should go in the mocks folder

ElliotFriedman commented on contracts/pcv/sentinel/guards/examples/ReEntrancyGuard.sol:
```diff
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
```
this should go in the mocks folder

ElliotFriedman commented on contracts/pcv/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,173 @@
+     */
+    function protec(address guard) 
+        external
+        override
```
this function should have a reentrancy guard

ditchfieldcaleb replied:
Now that we're not using delegatecall, and we're ensuring that we aren't calling ourselves with the address(this) check, no re-entrancy guard is needed

ElliotFriedman replied:
I would rather be safe than sorry, especially when executing arbitrary calldata against targets. We should add a reentrancy guard here.

ditchfieldcaleb replied:
There is literally no way that re-entrancy can happen. We're good.

thomas-waite commented on contracts/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,180 @@
+    {
+        for(uint256 i=0; i<whichGuards.length; i++) {
+            if (allowFailure) {
+                try this.protec(whichGuards[i]) {
```
Why not turn `protec` into a `public` function and just call `try protec(whichGuards[i])`? 

Not generally a fan of calling smart contract methods from within the contract itself using `this.` syntax

thomas-waite replied:
Ignore - needed to use `try`

thomas-waite reviewed (commented):
~Question about syntax~

ElliotFriedman commented on test/unit/sentinel/PCVSentinel.test.ts:
```diff
@@ -0,0 +1,216 @@
+// Import if needed, just a helper.
+// const toBN = ethers.BigNumber.from;
+
+describe.only('PCV Sentinel', function () {
```
remove .only

ElliotFriedman commented on test/unit/sentinel/PCVSentinel.test.ts:
```diff
@@ -0,0 +1,216 @@
+    });
+
+    it('reverts if condition not met', async () => {
+      await pcvSentinel.connect(impersonatedSigners[guardianAddress]).knight(balanceGuard.address);
```
add this into a beforeEach hook instead of duplicating this logic across multiple tests

ditchfieldcaleb replied:
Which ones we want knighted aren't the same across tests, otherwise I would

ElliotFriedman commented on contracts/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,180 @@
+            if (allowFailure) {
+                try this.protec(whichGuards[i]) {
+                    // the emit happens in protec
+                } catch {
```
If there is a failure, you should emit that error string and target guard smart contract so that it can be easily debugged

ditchfieldcaleb replied:
Good call, will do

ElliotFriedman commented on contracts/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,180 @@
+import "./IGuard.sol";
+
+/**
+ * @title PCV Sentinel
```
**style:** all comments should use `///` and not multi line comments to conform to our codebase

ditchfieldcaleb replied:
Both are natspec and should be acceptable imo

ElliotFriedman commented on contracts/sentinel/guards/examples/ReEntrancyGuard.sol:
```diff
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
```
Example contracts should be moved to the mocks folder

ditchfieldcaleb replied:
This isn't a mock, it need to be here so that people see it as an example.

ElliotFriedman reviewed (changes requested):
Things to be fixed before approval:
- Reentrancy guard on protec
- Natspec comments
- Enforce all or nothing execution on protecMany and remove `allowFailure` param
- Move examples to mocks folder

ElliotFriedman commented:
@ditchfieldcaleb lmk when you're ready for another review after changes have been pushed

ditchfieldcaleb commented on contracts/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,109 @@
+            for (uint256 i = 0; i < targets.length; i++) {
+                require(targets[i] != address(this), "Cannot target self.");
+                (bool success, ) = targets[i].call(calldatas[i]);
+                require(success, "Sub-call failed."); // Fail this call if any of the sub-calls fail
```
@TODO: pass through the error message

ElliotFriedman replied:
let's bubble up errors here if possible

ElliotFriedman replied:
bubble up error here

ElliotFriedman commented:
I'm really liking the new PCV Sentinel! It is very minimal and keeps things simple.

ElliotFriedman commented on contracts/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,109 @@
+    function knight(address guard)
+        external
+        override
+        isGovernorOrGuardianOrAdmin
```
Should a guardian be able to add a guard? This is essentially saying that guardian can add a contract into the system that can construct calldata that will be executed by a contract with PCV Controller role. It feels like this gives the guardians too much power.

ditchfieldcaleb replied:
@ElliotFriedman no that's not the case, where are you getting the idea that this is a PCVController still?

(It was, but shouldn't be anymore)

ditchfieldcaleb replied:
Just confirmed, yeah, this will use the PCVGuardian to pull funds to safety :)

ditchfieldcaleb replied:
Closing due to lack of further response, feel free to re-open @ElliotFriedman if you have further concerns.

ditchfieldcaleb commented:
> Things to be fixed before approval:
> 
> * Reentrancy guard on protec
> * Natspec comments
> * Enforce all or nothing execution on protecMany and remove `allowFailure` param
> * Move examples to mocks folder

- Reentrancy guard: added
- Examples aren't mocks. Yeah, perhaps it'd be better form to move them to the mocks folder, but practicality -> form, and in the mocks folder the examples would never be seen unless you already knew they existed. The examples are a reference for knowing how to create actual guards. Going to leave this as-is.
- Removed allowFailure & protecMany, vastly simplified
- Comments natspec'd.

ditchfieldcaleb commented:
@Joeysantoro @ElliotFriedman final review requested. I've made a major and potentially controversial change in the last commit: added receive() and fallback() as payable functions, and adjusted the Guard ABI to send an array of values along with the addresses and bytes/data array, which is used when calling the provided calldata.

Justification on receives() and payable(): any contracts that might incentivize calling them in eth. I don't think we do this now but decent chance we will at some point.

Justification on sending value: automation of non-critical amounts of pcv. Topping off small but necessary contracts or eoas. Sending funds received via payable() and receive() to some pre-determined place.

Flexible here, not super attached to the value in the calls, but fairly convinced that having receive() and fallback() is a good idea. Though if we remove the value in the calls we'll need to add in an admin method for recovering eth in the contract received via receive() and fallback().

ElliotFriedman commented on contracts/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,131 @@
+     * (in eth) from a contract as part of operation. For similar (and not unlikely) edge cases,
+     * the contract also has the capability of sending eth inside when instructed by a guard to do so.
+     */
+    receive() external payable {}
+
+    fallback() external payable {}
```
These two functions cannot exist as long as this contract is not a PCV deposit. If it's a PCV deposit then you can add them because you have a way to remove the eth that this contract could potentially receive. Right now, any eth that is received is locked forever.

Joeysantoro replied:
I think its fine

This contract isn't meant to hold meaningful amounts of ETH. Any "safe ETH" moves will be sent to DAO timelock

ElliotFriedman replied:
It also can specify how much eth to send in the `protec()` function so it's good to go at this point.

ElliotFriedman commented on contracts/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,131 @@
+            for (uint256 i = 0; i < targets.length; i++) {
+                require(targets[i] != address(this), "Cannot target self.");
+                (bool success, bytes memory returndata) = targets[i].call{
+                    value: values[i]
```
We should add a test that rescues ERC20 tokens and eth using this function and some rescue guards.

ElliotFriedman replied:
The rescue guards can be hardcoded to send to an immutable address, and in prod this can be the timelock.

ditchfieldcaleb replied:
Good call, adding

Joeysantoro commented on contracts/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,131 @@
+     * @param _core the Tribe core address
+     */
+    constructor(address _core) CoreRef(_core) {
+        _setContractAdminRole(keccak256("PCV_SENTINEL_ADMIN_ROLE"));
```
I think we should deprecate the "contract admin role" concept and use TribeRoles.

Joeysantoro commented on contracts/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,131 @@
+    function protec(address guard) external override nonReentrant {
+        require(guards.contains(guard), "Guard does not exist.");
+
+        if (IGuard(guard).check()) {
```
require(guard.check()) is cleaner than if-revert

Joeysantoro commented on contracts/sentinel/PCVSentinel.sol:
```diff
@@ -0,0 +1,131 @@
+    /**
+     * @notice returns a list of all guards
+     */
+    function allGuards() external view override returns (address[] memory all) {
```
I think we should add two methods for automation purposes:
1. GetAllProtecableGuards
2. protecAll

so the defender script can run protecAll(getAllProtecableGuards())

ditchfieldcaleb replied:
This can/will be implemented in a helper contract, no need to add it to the core API of the Sentinel

ditchfieldcaleb replied:
(esp because protecAll must necessarily be opinionated about how to handle reverts within single guards when running all of them, etc etc)

Joeysantoro reviewed (commented):
Left a few comments, none are dealbreakers but all are nice to have and worth it imo

Joeysantoro commented:
Love this contract btw. Attack surface is minimal and I believe it is prod-ready after the above changes are reviewed

Joeysantoro commented on contracts/sentinel/PCVSentinel.sol:
```diff
@@ -124,7 +116,7 @@ contract PCVSentinel is IPCVSentinel, CoreRef, ReentrancyGuard {
 
     /**
-     * @dev receive() and fallback() have been added and made payable for cases where the contract
+     * @dev receive() and fallback() have been added and made payable for cases wrhere the contract
```
typo
